pkg/agent/ipassigner/ip_assigner_linux.go

```diff
@@ -31,25 +32,188 @@ import (
 	"antrea.io/antrea/pkg/agent/util/arping"
 	"antrea.io/antrea/pkg/agent/util/ndp"
 	"antrea.io/antrea/pkg/agent/util/sysctl"
+	crdv1b1 "antrea.io/antrea/pkg/apis/crd/v1beta1"
 )
 
-// ipAssigner creates a dummy device and assigns IPs to it.
+// VLAN interfaces created by antrea-agent will be named with the prefix.
+// For example, when VLAN ID is 10, the name will be antrea-ext.10.
+// It can be used to determine whether it's safe to delete an interface when it's no longer used.
+const vlanInterfacePrefix = "antrea-ext."
+
+// assignee is the unit that IPs are assigned to. All IPs from the same subnet share an assignee.
+type assignee struct {
+	// logicalInterface is the interface IPs should be logically assigned to. It's also used for IP advertisement.
+	// The field must not be nil.
+	logicalInterface *net.Interface
+	// link is used for IP link management and IP address add/del operation. The field can be nil if IPs don't need to
+	// assigned to an interface physically.
+	link netlink.Link
+	// arpResponder is used for ARP responder for IPv4 address. The field should be nil if the interface can respond to
+	// ARP queries itself.
+	arpResponder responder.Responder
+	// ndpResponder is used for NDP responder for IPv6 address. The field should be nil if the interface can respond to
+	// NDP queries itself.
+	ndpResponder responder.Responder
+	// ips tracks IPs that have been assigned to this assignee.
+	ips sets.Set[string]
+}
+
+// deletable returns whether this assignee can be safely deleted.
+func (as *assignee) deletable() bool {
+	if as.ips.Len() > 0 {
+		return false
+	}
+	// It never has a real link.
+	if as.link == nil {
+		return false
+	}
+	// Do not delete non VLAN interfaces.
+	if _, ok := as.link.(*netlink.Vlan); !ok {
+		return false
+	}
+	// Do not delete VLAN interfaces not created by antrea-agent.
+	if !strings.HasPrefix(as.link.Attrs().Name, vlanInterfacePrefix) {
+		return false
+	}
+	return true
+}
+
+func (as *assignee) destroy() error {
+	if err := netlink.LinkDel(as.link); err != nil {
+		return fmt.Errorf("error deleting interface %v", as.link)
```

Comment:
not wrapping the actual error is intentional here?

Reply:
added